src/v/security/gssapi_principal_mapper.cc

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "security/gssapi_principal_mapper.h"
+
+#include "security/logger.h"
+#include "vassert.h"
+#include "vlog.h"
+
+#include <boost/algorithm/string/case_conv.hpp>
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+
+#include <charconv>
+
+namespace security {
+static constexpr const char* const non_simple_pattern = R"([/@])";
+static constexpr const char* const parameter_pattern = R"(([^$]*)(\$(\d*))?)";
+
+gssapi_rule::gssapi_rule(std::string_view default_realm)
+  : _default_realm(default_realm) {}
+
+gssapi_rule::gssapi_rule(
+  std::string_view default_realm,
+  int number_of_components,
+  std::string_view format,
+  std::string_view match,
+  std::string_view from_pattern,
+  std::string_view to_pattern,
+  repeat repeat_,
+  make_lower to_lower,
+  make_upper to_upper)
+  : _default_realm(default_realm)
+  , _is_default(false)
+  , _number_of_components(number_of_components)
+  , _format(format)
+  , _match(match)
+  , _from_pattern(std::regex{
+      from_pattern.begin(),
+      from_pattern.length(),
+      std::regex_constants::ECMAScript | std::regex_constants::optimize})
+  , _from_pattern_str(from_pattern)
+  , _to_pattern(to_pattern)
+  , _repeat(repeat_)
+  , _to_lower(to_lower)
+  , _to_upper(to_upper) {}
+
+std::optional<ss::sstring>
+gssapi_rule::apply(std::vector<std::string_view> params) const {
+    static const RE2 non_simple_regex(non_simple_pattern, RE2::Quiet);
```

Comment:
thread_local?

Reply:
Yeah good catch on that.

Side question -> when the rules are created (when the config binding is updated/set), is that tied to a specific shard?   I guess my question is, is the configuration stuff locked to a specific shard?

Reply:
> > where does this hook in for principal extraction? presumably that is part of the process that occurs after authentication when we map a connection to a user?
> 
> As per what @BenPope said. I started this PR right around when his got merged in. I was under the, possibly wrong, assumption that a separate PR would be required to piece the two together.
> 
> 1. Do you want me to do that in this PR?

I think either way is fine. 

> 2. I still have to add in the correct, I think 'features'(?) into the feature spec.

I don't think anything needs doing here.

> Also, I need to look a little bit more into what the source of "default realm" is. It appears to be a runtime generated item, not set by the user.

I think you'll get the canonical form of the principal as input; it should include the realm.

Reply:
>Side question -> when the rules are created (when the config binding is updated/set), is that tied to a specific shard? I guess my question is, is the configuration stuff locked to a specific shard?

yes, generally everything is locked to a shard. for this principle mapping stuff i would assume that it happens in the context of a connection, which is itself tied to a specific shard.